pkg/abstractions/endpoint/task.go

```diff
@@ -74,10 +74,15 @@ func (t *EndpointTask) Cancel(ctx context.Context, reason types.TaskCancellation
 }
 
 func (t *EndpointTask) HeartBeat(ctx context.Context) (bool, error) {
-	heartbeatKey := Keys.endpointRequestHeartbeat(t.msg.WorkspaceName, t.msg.StubId, t.msg.TaskId)
-	exists, err := t.es.rdb.Exists(ctx, heartbeatKey).Result()
+	task, err := t.es.backendRepo.GetTask(ctx, t.msg.TaskId)
+	if err != nil {
+		return false, err
+	}
+
+	key := Keys.endpointRequestHeartbeat(t.msg.WorkspaceName, t.msg.StubId, t.msg.TaskId, task.ContainerId)
```

Comment:
It should be safe to do this, because heartbeat is only called if the task is claimed (which means the task has an assigned container ID): ```
				heartbeat, err := task.HeartBeat(ctx)
				if err != nil {
					continue
				}
```